Automatic Documentation System for TypeScript Components

ashikmeerankutty opened the pull request:
### Summary

Fixes #3056 

- Replaced `react-docgen` with `react-docgen-typescript`
- Custom babel plugin to inject doc info to components

### Checklist

~~- [ ] Check against **all themes** for compatibility in both light and dark modes~~
~~- [ ] Checked in **mobile**~~
~~- [ ] Checked in **IE11** and **Firefox**~~
~~- [ ] Props have proper **autodocs**~~
~~- [ ] Added **[documentation](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md)** examples~~
~~- [ ] Added or updated **[jest tests](https://github.com/elastic/eui/blob/master/wiki/testing.md)**~~
~~- [ ] Checked for **breaking changes** and labeled appropriately~~
~~- [ ] Checked for **accessibility** including keyboard-only and screenreader modes~~
~~- [ ] A **[changelog](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md#changelog)** entry exists and is marked appropriately~~

### Verifying Props

 - [x]   Accordion
 - [x]   Bottom Bar
 - [x]   Flex
 - [x]   Flyout
 - [x]   Header
 - [x]   Horizontal Rule
 - [x]   Modal
 - [x]   Nav Drawer
 - [x]   Page
 - [x]   Panel
 - [x]   Popover
 - [x]   Resizable container
 - [x]   Spacer
 - [x]   Breadcrumbs
 - [x]   Button
 - [x]   Collapsible nav
 - [x]   Context Menu
 - [x]   Control Bar
 - [x]  Facet
 - [x]   Key Pad Menu
 - [x]  Link
 - [x]   Pagination
 - [x]   Tree View
 - [x]   Side Nav
 - [x]   Steps
 - [x]   Tabs
 - [x]   Data grid
 - [x]  Tables
 - [x]   In-memory tables
 - [x]   Aspect ratio
 - [x]   Avatar
 - [x]   Badge
 - [x]   Callout
 - [x]   Card
 - [x]   Code
 - [x]   Comment list
 - [x]   Description List
 - [x]   Drag And Drop
 - [x]   EmptyPrompt
 - [x]   Health
 - [x]   Icons
 - [x]   Image
 - [x]   List Group
 - [x]   Loading
 - [x]   Progress
 - [x]   Stat
 - [x]   Text
 - [x]   Title
 - [x]   Toast
 - [x]   ToolTip
 - [x]   Tour
 - [x]   Form controls
 - [x]   Form layouts
 - [x]   Compressed forms
 - [x]   Form validation
 - [x]   SuperSelect
 - [x]   Combo Box
 - [x]   Color Selection
 - [x]   Code Editor
 - [x]   Date Picker
 - [x]   Expression
 - [x]   Filter Group
 - [x]   EuiRange
 - [x]   EuiDualRange
 - [x]   EuiDualRange
 - [x]   EuiRangeTicks
 - [x]   EuiRangeLevels
 - [x]   EuiRangeLevels
 - [x]  Search Bar
 - [x]   Selectable
 - [x]   Suggest
 - [x]   Super Date Picker
 - [x]   Accessibility
 - [x]   Beacon
 - [x]   Color
 - [x]   Color Palettes
 - [x]   Context
 - [x]   Copy
 - [x]   CSS utility classes
 - [x]   Delay Hide
 - [x]   Delay Render
 - [x]   Error Boundary
 - [x]   Focus Trap
 - [x]   Highlight and mark
 - [x]   I18n
 - [x]   Outside Click Detector
 - [x]   Overlay Mask
 - [x]   Portal
 - [x]   ResizeObserver
 - [x]   Responsive
 - [x]   Toggle
 - [x]   Window Events

kibanamachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

thompsongl commented:
Nice work making progress on all the checklist items!
Is there anything else you'll need to do after you've checked all the components, or will this be ready for a formal review?

ashikmeerankutty commented:
I used that checklist for ensuring props are generated for all components. There are still some components that are missing some props. and also there are some issues regarding
- There is some issue with the **unchecked** components as the props for them are not generated
- **children prop**: if the children prop is given as required it's not shown as required by the parser also the type for children prop is found different.
- I need to omit some props that are both in `HtmlElement` and component prop definition as they are removed as the parser removes props from HtmlElement.

Rest all seems fine and I also feel that there are some types that are not much readable like 

![image](https://user-images.githubusercontent.com/10515204/85504649-123ae800-b60a-11ea-87da-3c8f2528d6b2.png)

ashikmeerankutty commented:
@thompsongl There seems to be some issues with the tests of `EuiLink` can you please check that. It seems like they are flaky

thompsongl commented:
> There is some issue with the unchecked components as the props for them are not generated

Just **EuiFlex**?


> children prop: if the children prop is given as required it's not shown as required by the parser also the type for children prop is found different.

Is it found different because `children` was not `Omit`-ed from the original HTML type?

ashikmeerankutty commented:
> Just **EuiFlex**?

The issue with `EuiFlex` is solved. Now props are generated for all the components although some props are missing some props.

> Is it found different because `children` was not `Omit`-ed from the original HTML type?
No, even if we omit the type from HTML type we are getting different outputs and the required is always false. When I dig in it was found to be an issue with `react-docgen-typescript` I will somehow try to fix it there (hopefully).

ashikmeerankutty commented:
Also as per a comment by @chandlerprall I have added a new property `extends` to every component in `docgenInfo` that can be used to show if a component is extending props from other modules 😃 .

thompsongl commented:
> There seems to be some issues with the tests of EuiLink

I think there's something else going on here because the props table for **EuiLink** is empty when I run it locally

ashikmeerankutty commented:
> I think there's something else going on here because the props table for **EuiLink** is empty when I run it locally

I have fixed that issue but I can't commit that. As the first test breaks even without a noticeable change

thompsongl commented:
> but I can't commit that

Go ahead and commit with the `-n` flag or comment out that test for now. No need for it to be a blocker at the moment.

snide commented:
jenkins test this

(for preview url to generate)

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3554/

ashikmeerankutty commented:
Overall progress with verifying  props are updated here 😄  :
https://www.notion.so/5abd1562413244d4ab292ab1251f1e49?v=71c0562f280b4807aeab2f6b83295cb3

thompsongl commented:
That is awesome and very helpful
(Also, I love Notion)

ashikmeerankutty commented:
I think this PR is ready for review as props are generated correctly for all the components. There are three issues that exist:
 - if the children prop is given as required it's not shown as required by the parser also the type for children prop is found different.
- Props of EuiOverlayProps as they are using `keyof HTMLElements` all the props with string type are displayed.
- Props default value for prop `responsive` was not generated.

Also some tests are broken I need to use `-n` flag for some commits. I think those are unrelated to this PR

ashikmeerankutty commented:
I think the issue with the children prop comes from the typescript. The same issue can be found in `intellisense` the prop of `children` is not taken from the user-defined type 😞 .

![image](https://user-images.githubusercontent.com/10515204/85953555-e3689d00-b98e-11ea-8704-f5a48799e8dc.png)

thompsongl commented:
> There are three issues that exist

We'll take a look and see if we can get these resolved or (at the very least) documented for understanding.

thompsongl commented:
> Props default value for prop responsive was not generated.

Which component?

ashikmeerankutty commented:
> Which component?

`EuiBreadCrumb`, the issue is that the default value is obtained from another variable.

thompsongl commented:
Let's get this branch up to date with master. I think diagnosing the test failures will be easier that way.

thompsongl commented:
PR for passing tests: ashikmeerankutty/eui#2

thompsongl commented:
jenkins test this

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3554/

thompsongl commented:
jenkins test this

flaky test

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3554/

thompsongl commented:
> if the children prop is given as required it's not shown as required by the parser also the type for children prop is found different.

My read is that `PropsWithChildren` is to blame here. `React.FunctionComponent` uses it and does not allow for an override of `children?`. I don't yet have an answer for it.

> Props of EuiOverlayProps as they are using keyof HTMLElements all the props with string type are displayed.

We may have to build a type that's used only for docs. This is a one-off case and I haven't found a way to capture it with the parser.

thompsongl commented:
@ashikmeerankutty See if you can find any more information on `PropsWithChildren`.

* Is there a way to override its application of `children`?
* Can the parser substitute a custom version of the type globally?

ashikmeerankutty commented:
> * Is there a way to override its application of `children`?

Found 3 ways to overcome the issue:

- By creating a custom interface removing `propsWithChildren` from `FunctionComponent`
like

```
    interface FunctionComponentWithoutChildren<P = {}> {
        (props: P, context?: any): ReactElement<any, any> | null;
        propTypes?: WeakValidationMap<P>;
        contextTypes?: ValidationMap<any>;
        defaultProps?: Partial<P>;
        displayName?: string;
    }
```
It works fine. but doesn't seem good.

- The next method I found is a lot easier right inside the plugin we could call for a function whenever a `children` prop is encountered and matches `children` prop from the source file itself using `regex` and replace `children prop` in `docgeInfo`.  I tried this and implemented and committed. It works fine. But I have no idea If it creates any other issue.

- The next method is to find `children prop` from the `AST` whenever `children prop` is encountered. It seems doable.

Which one would be better ?

thompsongl commented:
Thoughts on the `children` approaches, @chandlerprall?

chandlerprall commented:
A little hesitant around the regex approach - reading the code it looks like the childrenProp state is set & processed per-file, which means if we have two components in one file, I believe it will be copied & processed for both?

If possible & clean, the AST approach seems safer to me to avoid problems with that type of over-processing.

ashikmeerankutty commented:
> If possible & clean, the AST approach seems safer to me to avoid problems with that type of over-processing.

I have no idea how this ends up. I'm trying this.

thompsongl commented:
> I have no idea how this ends up. I'm trying this.

Anything you want help with or need input on, @ashikmeerankutty?

ashikmeerankutty commented:
> Anything you want help with or need input on, @ashikmeerankutty?

I figured out the type of children from `AST`. There are a few issues still. I will comment here after committing the changes.

thompsongl commented:
> I figured out the type of children from AST. There are a few issues still. I will comment here after committing the changes.

Nice! Sounds good

ashikmeerankutty commented:
Found `children` prop types for some of the functional components. There are a few issues that need to be fixed:
 - Types have to be parsed to string in a better way. ( I hope I could reuse the method in `prop-types-to-ts-prop` that looks soo good 😃  )
-  There are some issues with `children` props of class components too. I don't know why. 
-  Prop types are not found correctly if the children prop comes from a referenced `type` or `interface`

ashikmeerankutty commented:
> * Types have to be parsed to string in a better way. ( I hope I could reuse the method in `prop-types-to-ts-prop` that looks soo good 😃  )

Fixed by obtaining types from typescript itself

> * Prop types are not found correctly if the children prop comes from a referenced `type` or `interface`

Fixed

ashikmeerankutty commented:
> * There are some issues with `children` props of class components too. I don't know why.

Fixed

thompsongl commented:
Ready for review, @ashikmeerankutty?
If so, get this branch up to date with master and we'll start going through the code.

ashikmeerankutty commented:
> Ready for review, @ashikmeerankutty?

Yes

thompsongl commented:
There are a couple output differences from what the new playground component expects from `docgenInfo`. They are slight, so I'll open a PR to get it fixed.

I also noticed that `EuiFlexItem` shows the `component` type as a very long list of HTML elements. This is technically correct, but we should find some other way to indicate it. Like "any HTML element", or something. Any ideas here, @chandlerprall?

If you want to start in on code review, too, @chandlerprall, that'd be great.

thompsongl commented:
`EuiAvatar` `initialsLength` does not get typed as an enum, even though it is. It shows as `type: {name: "1 | 2"}`, and my guess is that numbers (instead of strings) have something to do with it.

ashikmeerankutty commented:
> `EuiAvatar` `initialsLength` does not get typed as an enum, even though it is. It shows as `type: {name: "1 | 2"}`, and my guess is that numbers (instead of strings) have something to do with it.

I guess its because it is of union type. There is an option in `react-docgen-typescript` to expand union types so that it fixes this issue. But it creates some more issues as it expands even `boolean` type to `true` or `false` similarly for a lot of ther props.

chandlerprall commented:
> I also noticed that EuiFlexItem shows the component type as a very long list of HTML elements. This is technically correct, but we should find some other way to indicate it. Like "any HTML element", or something. Any ideas here, @chandlerprall?

Your suggestion sounds right, if we can some-how intercept the `keyof JSX.IntrinsicElements` type and list it like "an HTML element" that'd be great.

ashikmeerankutty commented:
> Your suggestion sounds right, if we can some-how intercept the `keyof JSX.IntrinsicElements` type and list it like "an HTML element" that'd be great.

From the `propFilter` I could change the name of a prop. But I have no idea what condition I could use. I could use one like if `props.name === expanded keyof JSX.IntrinsicElements` Is that fine ?

thompsongl commented:
> I could use one like if props.name === expanded keyof JSX.IntrinsicElements Is that fine ?

Sounds reasonable for this case

thompsongl commented:
PR to reconcile playground input and docgen output: ashikmeerankutty/eui#3

ashikmeerankutty commented:
> Sounds reasonable for this case

I have committed the change can you please check if its okay.

thompsongl commented:
jenkins test this

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3554/

chandlerprall commented:
Just a note that I'm starting a deep review of these changes.

thompsongl commented:
> I have committed the change can you please check if its okay.

<img width="315" alt="Screen Shot 2020-07-14 at 3 31 43 PM" src="https://user-images.githubusercontent.com/2728212/87482858-28274180-c5e7-11ea-8e4c-73616c942ba2.png">

Looks good to me!

thompsongl commented on scripts/babel/react-docgen-typescript.js:
```diff
@@ -0,0 +1,282 @@
+
+  if (prop.type.name === 'enum') {
+    const propValueArray = prop.type.value.map(type => type.value);
+    const found = intrinsicValuesRaw.some(
```
Probably need to be more strict than `.some` here. `a` is in the list, which is used in several cases of `a | button` where not all HTML elements are allowed.

ashikmeerankutty replied:
Fixed it by replacing some with every

thompsongl commented:
jenkins test this

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3554/

chandlerprall commented on scripts/babel/react-docgen-typescript.js:
```diff
@@ -0,0 +1,282 @@
+ */
+/* eslint-disable @typescript-eslint/no-var-requires */
+
+const propsParser = require('react-docgen-typescript');
```
Only comment on this file is it'd be great to have some more comments detailing _why_ the code is doing a certain thing. Some limited examples:

* What does `intrinsicValuesRaw` represent
* Why return if `moduleSymbol` is not defined

ashikmeerankutty replied:
👍

ashikmeerankutty replied:
`!moduleSymbol` return was un necessary. As I am only taking typescript components. I added it while testing, forgot to avoid.

ashikmeerankutty commented on .babelrc.js:
```diff
@@ -23,14 +23,8 @@ module.exports = {
     "@babel/plugin-syntax-dynamic-import",
     "pegjs-inline-precompile",
-    "./scripts/babel/proptypes-from-ts-props",
+    "./scripts/babel/react-docgen-typescript",
```
No it can be deleted

ashikmeerankutty commented on src/components/accessibility/keyboard_accessible.ts:
```diff
@@ -44,6 +44,9 @@ import { Component, cloneElement, KeyboardEvent, ReactElement } from 'react';
+  /**
+   * ReactNode to render as this component's children
+   */
   children: ReactElement<any>;
```
👍

ashikmeerankutty replied:
Should I change that in the source file or in the parser itself

ashikmeerankutty replied:
Its a weird issue. In some case ReactNode is rendered as ReactNode and some case as a large string. I will use some conditions to solve this issues in the plugin itself.

chandlerprall replied:
I still see `EuiKeyboardAccessible`'s type as the expanded list

ashikmeerankutty commented on src/components/form/field_password/field_password.tsx:
```diff
@@ -47,6 +50,7 @@ export type EuiFieldPasswordProps = InputHTMLAttributes<HTMLInputElement> &
      * `string` | `ReactElement` or an array of these
      */
     append?: EuiFormControlLayoutProps['append'];
+    value?: string | string[] | number;
```
I got the type fo value from the types of `htmlElement`. I will change that

chandlerprall replied:
It now renders as a `ReactText`, which is defined as `string | number` so it isn't _wrong_, but I find it weird that something is mapping it that way. Mind looking into what is causing the indirection?

ashikmeerankutty replied:
I will check that

ashikmeerankutty replied:
I couldn't figure it out. whenever there is a string | number it is taking as ReactText

ashikmeerankutty commented on src/components/badge/badge_group/badge_group.tsx:
```diff
@@ -42,7 +42,7 @@ export interface EuiBadgeGroupProps {
 }
 
-export const EuiBadgeGroup = React.forwardRef<
+export const EuiBadgeGroup = forwardRef<
```
👍

ashikmeerankutty replied:
👍 . For that I think its better to add a feature in `react-docgen-typescript` itself. Is that ok if I try it there ?

chandlerprall replied:
Yeah, go for it! Either way, doesn't need to be part of this iteration on the props.

ashikmeerankutty commented:
> I assume dropping the `children` prop in a number of places is intentional?

I missed it 😞

ashikmeerankutty commented:
> When starting the dev server, webpack emits a bunch of warnings around exporting values which don't exist - because they are types/interfaces that are removed at TS->JS transform

#3771 Solves this issue partially as the removed types and interfaces are converted and appended as objects in the same file 😄 . Only those types imported from external packages shows up.

ashikmeerankutty commented:
> It could be nice to detect these code/typescript types

I think its possible using some regex patterns or a hash set. Which one would be better or is there any better way ?

miukimiu commented:
Hi @ashikmeerankutty, @chandlerprall and @thompsongl,  

I was taking a look at this PR and I'm wondering if there is a way to also automatically document the structure of option objects.

For example, the combo box `options` type is `EuiComboBoxOptionOption<T>[]`. But it doesn't list the available props of the object. There's an issue for this (https://github.com/elastic/eui/issues/1981).

For example, in the color palette picker component we're exposing the objects manually:
<img width="944" alt="Screenshot 2020-07-27 at 13 22 01" src="https://user-images.githubusercontent.com/2750668/88541166-35432800-d00c-11ea-956e-08201228d833.png">

To do this we're creating components to expose those props and then we're calling them in our docs:

```js
props: {
  EuiColorPalettePicker,
  EuiColorPalettePickerPaletteText,
  EuiColorPalettePickerPaletteFixed,
  EuiColorPalettePickerPaletteGradient,
}
```

Is this something we need to do to all the option objects or is there a way to automatize?

ashikmeerankutty commented:
> To do this we're creating components to expose those props and then we're calling them in our docs:

#3711 Allows passing types and interfaces directly to the docs. So that no components has to be created. 

For complete automation I guess something like combining #3711 and #3544 could work so that docgenInfo also could contain all the interfaces used in that file and we could create something like a tooltip to show the props of the interface or a link.

thompsongl commented:
> #3711 Allows passing types and interfaces directly to the docs. So that no components has to be created.

#3771 is the correct one 😄 

But, yes, @miukimiu, that functionality is coming soon.

chandlerprall commented on src/components/accessibility/screen_reader.tsx:
```diff
@@ -21,6 +21,9 @@ import { cloneElement, ReactElement, FunctionComponent } from 'react';
 
 export interface EuiScreenReaderOnlyProps {
+  /**
+   * ReactNode to render as this component's content
```
```suggestion
   * ReactElement to render as this component's content
```

chandlerprall commented:
> #3771 Solves this issue partially as the removed types and interfaces are converted and appended as objects in the same file 😄 . Only those types imported from external packages shows up.

Not the intentionally complicate things further, I created #3801 in a desire to make multiple babel configurations. Specifically, to remove docgen processing & code from our built npm distribution. This would mean not generating those intermediate interface/type components, and would re-trigger the webpack errors in downstream project 😞

chandlerprall commented:
> > It could be nice to detect these code/typescript types
> 
> I think its possible using some regex patterns or a hash set. Which one would be better or is there any better way ?

Maybe even the presence of `(` and `)` - maybe also `=>`? The only other way the doc's type would have those character (I think) is an enum of strings that happens to overlap?

chandlerprall commented:
(also wanted to say ❤️ thank you again for all the work & changes on this)

ashikmeerankutty commented:
> Not the intentionally complicate things further, I created #3801 in a desire to make multiple babel configurations. Specifically, to remove docgen processing & code from our built npm distribution. This would mean not generating those intermediate interface/type components, and would re-trigger the webpack errors in downstream project 😞

For suppressing the warnings this method can be used https://github.com/TypeStrong/ts-loader/issues/653#issuecomment-390889335. Is it okay including this ?

chandlerprall commented:
> For suppressing the warnings this method can be used TypeStrong/ts-loader#653 (comment). Is it okay including this ?

Unfortunately that would need to be used in downstream projects as well - which we'd like to avoid - as the code itself still points to non-existent exports.

chandlerprall commented:
The previous custom babel plugin for generating these types included a path that removed exports that pointed at  types/interfaces. Maybe this logic/approach be added to the new `react-docgen-typescript` plugin? https://github.com/elastic/eui/blob/0fc7625c12abcd2242ba5da354431edabe5af387/scripts/babel/proptypes-from-ts-props/index.js#L1426-L1463

A needed change would be in determining if an exported identity that comes from another file is a TS type. The existing plugin resolves all available types (imported & defined in the file) into `typeDefinitions`, which allows a simple lookup by the time the export-removal code executes. Does the TypeScript service / program provide similar information?

ashikmeerankutty commented:
> Does the TypeScript service / program provide similar information?

I will check that out.

ashikmeerankutty commented:
> The previous custom babel plugin for generating these types included a path that removed exports that pointed at types/interfaces. Maybe this logic/approach be added to the new `react-docgen-typescript` plugin?
> 
> https://github.com/elastic/eui/blob/0fc7625c12abcd2242ba5da354431edabe5af387/scripts/babel/proptypes-from-ts-props/index.js#L1426-L1463
> 
> A needed change would be in determining if an exported identity that comes from another file is a TS type. The existing plugin resolves all available types (imported & defined in the file) into `typeDefinitions`, which allows a simple lookup by the time the export-removal code executes. Does the TypeScript service / program provide similar information?

Warnings are removed. Can you please review it 😃

ashikmeerankutty commented:
### Works pending

> Maybe even the presence of `(` and `)` - maybe also `=>`? The only other way the doc's type would have those character (I think) is an enum of strings that happens to overlap?

I'm currently working on this. I am able to find functions using a regex. Currently experimenting the best way to show that in docs with some text highlighting.

> In a follow-up, it might be nice to sort the prop tables in some way to keep entries predictable.

I think sorting props by name is the best way. Is there any other suggestions? What would be better sorting inside plugin or in the docs page ?

> Idea for a follow-up: include `ref` in the props when a component uses `forwardRef` and identify what value it represents - either detected programmatically or via a comment

I haven't started it. Hope it can be completed soon 😅 

Also the issue with string | number transformed to ReactText. I couldn't figure out why till now

Is there anything else I'm missing 🤔  .

chandlerprall commented:
Looks like the merge with master branch broke something

> playground.js:16 Uncaught TypeError: Cannot read property 'custom' of undefined
    at facetButtonConfig (playground.js:16)
    at routes.js:277
    at Array.map (<anonymous>)
    at createExample (routes.js:276)
    at routes.js:357
    at Array.map (<anonymous>)
    at Module../routes.js (routes.js:343)
    at __webpack_require__ (bootstrap:769)
    at fn (bootstrap:129)
    at Module../store/configure_store.js (configure_store.js:1)

ashikmeerankutty commented:
> Looks like the merge with master branch broke something

There is an issue with EuiLink props. when I fix that prettier throws an error. Can you please check

chandlerprall commented:
Tested the removal of types/interface exports using commit https://github.com/elastic/eui/pull/3554/commits/54b3d5803b13fd3de7bd61f32a971fcc17a4dc4f - looks good!

> I think sorting props by name is the best way. Is there any other suggestions? What would be better sorting inside plugin or in the docs page ?

Let's sort in the docs page. I think that gives a bit more obvious place if we want more flexibility in the future

> Is there anything else I'm missing 🤔 .

I can't think of anything you've missed. @thompsongl that list in https://github.com/elastic/eui/pull/3554#issuecomment-670576426 look good to you?

chandlerprall commented:
> There is an issue with EuiLink props. when I fix that prettier throws an error. Can you please check

I appear to have success by changing it to `...(propsToUse.onClick || {}).custom,`

ashikmeerankutty commented:
> I appear to have success by changing it to `...(propsToUse.onClick || {}).custom,`

There is another issue that onClick Prop is not generated by the plugin due to the usage of

```
const EuiLink = forwardRef<HTMLAnchorElement | HTMLButtonElement, EuiLinkProps>(
```

here HTMLAnchorElement | HTMLButtonElement results in onClick props from 'DOM Attributes' so the `onClick` prop will not be generated. 

We already use the in `EuiLinkButtonProps` and `EuiLinkAnchorProps` where I omitted `onClick` prop from 'DomAttributes'.
but when I try to remove them inside forwardRef prettier starts throwing error.

ashikmeerankutty commented:
Fixed 😃

chandlerprall commented:
Ah, good catch; makes sense after seeing the required change.

I think the `React.ReactText` thing is the only piece blocking this PR from merging, can you prioritize that? The rest of the list is additional nice-to-have functionality that can be done separately. Would like to get this in before making it even more complex 😄

thompsongl commented:
> I can't think of anything you've missed. @thompsongl that list in #3554 (comment) look good to you?

Yes, looks good to me

ashikmeerankutty commented:
> I think the `React.ReactText` thing is the only piece blocking this PR from merging, can you prioritize that?

The thing with ReactText is weird. If HTMLAttributes are included it will show as ReactText and If not it will show as string | number. The simples fix I can think of is to return "string | number" as props if the type is ReactText.

thompsongl commented:
jenkins test this

ashikmeerankutty commented:
👀

ashikmeerankutty commented:
@thompsongl Fixed issue with the links. When I try to commit this error seems coming and its totally unrelated so I had to use -n flag

![image](https://user-images.githubusercontent.com/10515204/89937874-4bb7c900-dc33-11ea-843e-f6a12b85197e.png)

thompsongl commented:
I do get some `prettier` errors when running `yarn test` locally. We'll see what CI comes back with

jenkins test this

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3554/

thompsongl commented:
Yep, 3 valid `prettier` errors that will be easy to fix. Try merging master and running `yarn` again, too, to solve the TS error.

ashikmeerankutty commented:
> Yep, 3 valid `prettier` errors that will be easy to fix. Try merging master and running `yarn` again, too, to solve the TS error.

I Fixed the errors in the jenkins output. Can you please run test again.

thompsongl commented:
jenkins test this

chandlerprall commented:
> The simplest fix I can think of is to return "string | number" as props if the type is ReactText.

Let's do that. I've never encountered ReactText before, and is likely to be confusing in more situations than it resolves.

ashikmeerankutty commented:
> Let's do that. I've never encountered ReactText before, and is likely to be confusing in more situations than it resolves.

I have already done that 🚀

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3554/

chandlerprall commented:
`EuiFieldPassword`'s props look good!

Remember when I asked you to remove the previous `./scripts/babel/proptypes-from-ts-props` plugin? Turns out we still want to keep generating proptypes for our components 😅 Can you please re-add that plugin file & entry in the babel configuration? I don't expect it to conflict with the other changes here, but please let me know if it causes trouble and I'll help clean it up.

Sorry about that 🤦, I expect it to be the last piece of this PR! 🎉

ashikmeerankutty commented:
> I don't expect it to conflict with the other changes here, but please let me know if it causes trouble and I'll help clean it up.

There aren't any conflicts. But the test keeps failing since there is new lines are appended to the source file before 
```
FooComponent.PropTypes = ...
```
and after 
```
import PropTypes from 'proptypes'
```

thompsongl commented:
I'll look at fixing the tests, @chandlerprall, if you haven't started yet

thompsongl commented:
jenkins test this

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3554/

thompsongl commented:
One last time

jenkins test this

thompsongl commented:
> JavaScript heap out of memory

jenkins test this

thompsongl commented:
hmmm

> JavaScript heap out of memory

again. Checking on whether something's up with Jenkins

thompsongl commented:
> JavaScript heap out of memory

happening locally, also. Looking into increasing memory limit

thompsongl commented:
jenkins test this

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3554/

thompsongl commented:
Needed to increase the limit on the axe script, also. As such, I'm going to start work on #3801, but we can still move forward here and not block merging.

jenkins test this

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3554/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3554/
